crates/modelardb_server/src/optimizer/model_simple_aggregates.rs

```diff
@@ -21,122 +26,186 @@ use std::sync::Arc;
+) -> Result<Arc<dyn ExecutionPlan>> {
+    // aggregate_exec.input_schema().fields.len() == 1 should prevent this, but SortedJoinExec can
+    // be constructed with multiple inputs so the number of children is checked just to be sure.
+    if grid_execs.len() > 1 {
+        return Err(DataFusionError::NotImplemented(
+            "All aggregates must be for the same FIELD column.".to_owned(),
+        ));
+    }
+
+    // A GridExec can only have one child, so it is not necessary to check it.
+    let grid_exec_child = grid_execs[0].children().remove(0);
+    if let Some(parquet_exec) = grid_exec_child.as_any().downcast_ref::<ParquetExec>() {
+        if parquet_exec.predicate().is_some() || parquet_exec.pruning_predicate().is_some() {
+            return Err(DataFusionError::NotImplemented(
+                "Predicates cannot be pushed before the aggregate.".to_owned(),
+            ));
+        }
+    } else {
+        return Err(DataFusionError::Plan(
+            "The input to GridExec must be a ParquetExec.".to_owned(),
+        ));
+    }
+
+    let model_based_aggregate_exprs = try_rewrite_aggregate_exprs(aggregate_exec)?;
+
+    // unwrap() is safe as the input is from the existing AggregateExec.
+    Ok(Arc::new(
+        AggregateExec::try_new(
+            *aggregate_exec.mode(),
+            aggregate_exec.group_expr().clone(),
+            model_based_aggregate_exprs,
+            aggregate_exec.filter_expr().to_vec(),
+            grid_execs[0].children().remove(0),
+            aggregate_exec.schema(),
+        )
+        .unwrap(),
+    ))
+}
+
+/// Return [`AggregateExprs`](AggregateExpr) that computes the same aggregates as `aggregate_exec`
+/// if the aggregates are all computed for the same FIELD column, otherwise [`DataFusionError`] is
+/// returned.
+fn try_rewrite_aggregate_exprs(
+    aggregate_exec: &AggregateExec,
+) -> Result<Vec<Arc<dyn AggregateExpr>>> {
+    let aggregate_exprs = aggregate_exec.aggr_expr();
+    let mut rewritten_aggregate_exprs: Vec<Arc<dyn AggregateExpr>> =
+        Vec::with_capacity(aggregate_exprs.len());
+
+    for aggregate_expr in aggregate_exprs {
+        if aggregate_expr.as_any().is::<Count>() {
+            rewritten_aggregate_exprs.push(ModelAggregateExpr::new(ModelAggregateType::Count));
+        } else if aggregate_expr.as_any().is::<Min>() {
+            rewritten_aggregate_exprs.push(ModelAggregateExpr::new(ModelAggregateType::Min));
+        } else if aggregate_expr.as_any().is::<Max>() {
+            rewritten_aggregate_exprs.push(ModelAggregateExpr::new(ModelAggregateType::Max));
+        } else if aggregate_expr.as_any().is::<Sum>() {
+            rewritten_aggregate_exprs.push(ModelAggregateExpr::new(ModelAggregateType::Sum));
+        } else if aggregate_expr.as_any().is::<Avg>() {
+            rewritten_aggregate_exprs.push(ModelAggregateExpr::new(ModelAggregateType::Avg));
+        } else {
+            return Err(DataFusionError::Internal(format!(
```

Comment:
Is it necessary to raise an error? Can't the unoptimized aggregate just be used?

Reply:
The error is returned to `try_new_aggregate_exec()` on Line 160, which returns it to `rewrite_agggregates_to_use_segments()` on Line 117, which means the `if let Ok(input) = ...` is not `true`, so execution simply continues to the end of `rewrite_agggregates_to_use_segments()` where the original physical query plan is returned.

Reply:
OK